### PR TITLE
CDAP-1150: Feature/preferences cli

### PR DESCRIPTION
**Pull request by gokulavasan**
Summary :

* Integrating PreferencesClient into CDAP CLI.
* Adding a end-to-end test in PreferencesClientTest to test the programmatic API changes (ie, to verify if preferences do indeed get passed to the program via ProgramContext).

JIRA : https://issues.cask.co/browse/CDAP-1150

Bamboo: https://builds.cask.co/browse/CDAP-DUT488

### CODE REVIEW

File: cdap-cli/src/main/java/co/cask/cdap/cli/command/AbstractGetPreferencesCommand.java [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+import com.google.gson.Gson;
+
+import java.io.PrintStream;
+
+/**
+ * Abstract Class for getting preferences for instance, namespace, application, program.
+ */
+public abstract class AbstractGetPreferencesCommand extends AbstractAuthCommand {
+  private static final Gson GSON = new Gson();
+  private static final String DEFAULT_NAMESPACE = Constants.DEFAULT_NAMESPACE;
```
Comment (bdmogal) [inline cdap-cli/src/main/java/co/cask/cdap/cli/command/AbstractGetPreferencesCommand.java]: Why create another constant?

File: cdap-cli/src/main/java/co/cask/cdap/cli/command/GetResolvedPreferencesCommand.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+  private final ElementType type;
+
+  protected GetResolvedPreferencesCommand(ElementType type, PreferencesClient client, CLIConfig cliConfig) {
+    super(type, client, cliConfig, true);
+    this.type = type;
+  }
+
+  @Override
+  public String getPattern() {
+    return String.format("get %s preferencesresolved [<%s>]", type.getName(), type.getArgumentName());
```
Comment (bdmogal) [inline cdap-cli/src/main/java/co/cask/cdap/cli/command/GetResolvedPreferencesCommand.java]: ``get %s resolved preferences``?

File: cdap-client-tests/src/test/java/co/cask/cdap/client/PreferencesClientTestRun.java [hunk 3]
```diff
@@ -48,10 +62,64 @@ public void setUp() throws Throwable {
     super.setUp();
     client = new PreferencesClient(clientConfig);
     appClient = new ApplicationClient(clientConfig);
+    serviceClient = new ServiceClient(clientConfig);
+    programClient = new ProgramClient(clientConfig);
     namespaceClient = new NamespaceClient(clientConfig);
   }
 
   @Test
+  public void testProgramAPI() throws Exception {
```
Comment (bdmogal) [inline cdap-client-tests/src/test/java/co/cask/cdap/client/PreferencesClientTestRun.java]: We would need to add namespace-related logic here later right? Could you please add a comment?

**Comment by bdmogal** [comment]
Could you report a jira to add the namespace-related logic in this CLI later?

**Comment by bdmogal** [comment]
Minor comments, :+1: otherwise

File: cdap-client-tests/src/test/java/co/cask/cdap/client/PreferencesClientTestRun.java (hunk 3 above)
Comment (gokulavasan) [inline cdap-client-tests/src/test/java/co/cask/cdap/client/PreferencesClientTestRun.java]: Added

**Pull request merged**